File: scripts/mechanical_gate.ts
/**
 * Mechanical Gate (TypeScript Implementation)
 */

interface ExecutionIntent {
    action: string;
    target?: string;
    query?: string;
    [key: string]: any;
}

export class MechanicalGate {
    private readonly blacklistPatterns: RegExp[] = [
        /rm\s+-rf\s+\//i,
        /mkfs/i,
        />\s*\/dev\/sda/i,
        /DROP\s+TABLE/i,
        /DELETE\s+MATCH/i
    ];

    public validateCommand(commandIntent: string): boolean {
        for (const pattern of this.blacklistPatterns) {
            if (pattern.test(commandIntent)) {
                console.error(`[MECHANICAL_GATE_BLOCK] Destructive pattern detected: ${pattern}`);
                return false;
            }
        }
        return true;
    }

    public cerebellumTranslate(intentJson: string): string {
        try {
            const data: ExecutionIntent = JSON.parse(intentJson);
            if (data.action === "query_graph" && data.query) {
                return `cypher: ${data.query}`;
            }
            return `Translated execution for: ${data.action}`;
        } catch (error) {
            console.error("[MECHANICAL_GATE_ERROR] Invalid Intent JSON structure.");
            return "";
        }
    }
}

File: scripts/mechanical_gate.py
import re
import sys
import json

class MechanicalGate:
    """
    Non-LLM deterministic guardrail.
    Intercepts [DO] execution commands from the LLM and runs them through safety filters.
    """
    def __init__(self):
        self.blacklist_patterns = [
            r"rm\s+-rf\s+/",
            r"mkfs",
            r">\s*/dev/sda",
            r"DROP\s+TABLE",
            r"DELETE\s+MATCH"
        ]

    def validate_command(self, command_intent: str) -> bool:
        for pattern in self.blacklist_patterns:
            if re.search(pattern, command_intent, re.IGNORECASE):
                print(f"[MECHANICAL_GATE_BLOCK] Destructive pattern detected: {pattern}")
                return False
        return True

    def cerebellum_translate(self, intent_json: str) -> str:
        try:
            data = json.loads(intent_json)
            if data.get("action") == "query_graph":
                return f"cypher: {data.get('query')}"
            return f"Translated execution for: {data.get('action')}"
        except json.JSONDecodeError:
            print("[MECHANICAL_GATE_ERROR] Invalid Intent JSON.")
            return ""

if __name__ == "__main__":
    gate = MechanicalGate()
    if len(sys.argv) > 1:
        test_cmd = sys.argv[1]
        is_safe = gate.validate_command(test_cmd)
        print(f"Command Safe: {is_safe}")
